designdocs/CAPACITY.md

```diff
@@ -0,0 +1,147 @@
+
+Once staked,  Capacity is granted at a ratio  6,000 token to 1 unit of capacity.  This can then be used to pay for transactions. 
+
+**Below** **are transactions that can be paid with Capacity:**
+
+- Create an MSA
+- Add a key to an MSA
+- Delegate permissions to another MSA
+- Update Delegate Permissions
+- Send a message
+- Send a batch message
+
+As mentioned above, Capacity is non-transferable and at a minimum implements the following interface.
+
+```rust
+traits Nontransferable {
+   type Balance = Balance;
+
+   fn available(msa_id: MessageSenderId) -> Result<Balance, DispatchError>;
+   fn reduce_available(msa_id: MessageSenderId, amount: Balance) -> Result<Balance, DispatchError>;
+   fn increase_available(msa_id: MessageSenderId, amount: Balance) -> Result<Balance, DispatchError>;
+ }
+
+```
+
+We can represent Capacity as a mapping of an MsaId to information containing the total balance available to an MSA.
+
+```rust
+#[pallet::storage]
+pub type CapacityOf<T: Config> = StorageMap<_, Blake2_128Concat, MessageSenderId, Details<T::Balance>>;
+
+pub struct Details {
+	pub used: Balance,
+	pub available: Balance,
+}
+```
+
+And we can store the a token account staking to MSA by mapping.
+
+```rust
+#[pallet::storage]
+pub type Staking<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, StakingDetails<T::Balance>>;
+
+pub struct StakingDetails<Balance> {
+	pub msa_id: MessageSenderId,
+	pub total_reserved: Balance,
+}
+```
+
+To give the limitations of what you can do with Capacity you are limited to staking and un-staking Capacity for token. Note that these interface can change as we introduce stash and controller accounts using Proxy pallet.
+
+```rust
+pub fn stake(origin: OriginFor<T>, account: MessageSenderId, amount: BalanceOf<T>)
+
+pub fn stake(origin: OriginFor<T>, account: MessageSenderId, amount: BalanceOf<T>)
+```
+
+Events for triggers for staking include
+
+```rust
+pub enum Event<T: Config> {
+  Staked { account: T::AccountId, msa_id: MessageSenderId, amount: BalanceOf<T> },
+
+  UnStake{ account: T::AccountId, msa_id: MessageSenderId, amount: BalanceOf<T> },
+}
+```
+
+Errors when staking
+
+```rust
+pub enum Error<T> {
+	AlreadyStaked,
+	InsufficientBalance,		
+	InvalidMsa,
+}
+```
+
+**Capacity is replenishable**
+
+Replenishable mean that after a certain fixed period, called an epoch, all capacity is replenished. The epoch period is composed of  blocks. A block is filled with 25% transactions being operational and 75% for all other transactions including Capacity. This can be configurable by setting the  `AvailableBlockRatio` . Moreover, one unit of Capacity is `10^12` and matches with substrate weight measurement where `10^12` is 1 second of execution time.
```

Comment:
add note about how in the beginning we will be having a fix epoch length that is configurable. At some point it will evolve to something like this.